internal/impl/timeplus/output.go

```diff
@@ -0,0 +1,216 @@
+
+This output accepts structured message only. It also expects all message contains the same keys and matches the schema of stream. If the upstream source returns
+unstructured message such as string, please include a processors to contruct the strcuture message. Here is a short sample:
+
+` + "```yml" + `
+output:
+  timeplus:
+    workspace: my_workspace_id
+    stream: mystream
+    apikey: <Api key genereated on web console>
+
+	processors:
+	  - mapping: |
+        root = {}
+        root.raw = content().string() # Make sure your stream contains a string type raw field
+` + "```" + `
+
+A sample config to send the data to Timeplus Enterprise Cloud
+` + "```yml" + `
+output:
+  timeplus:
+    workspace: my_workspace_id
+    stream: mystream
+    apikey: <Api key genereated on web console>
+` + "```" + `
+
+A sample config to send the data to Timeplus Enterprise (self-hosted)
+` + "```yml" + `
+output:
+  timeplus:
+    url: http://localhost:8000
+    workspace: my_workspace_id
+    stream: mystream
+    username: username
+    password: pw
+` + "```" + `
+
+A sample config to send the data to timeplusd
+` + "```yml" + `
+output:
+  timeplus:
+    url: http://localhost:3218
+    stream: mystream
+    username: username
+    password: pw
+` + "```" + `
+`)
+
+	outputConfigSpec.
+		Field(service.NewStringEnumField("target", http.TargetTimeplus, http.TargetTimeplusd).Default(http.TargetTimeplus).Description("The destination type, either Timeplus Enterprise or timeplusd")).
+		Field(service.NewURLField("url").Description("The url should always include schema and host.").Default("https://us-west-2.timeplus.cloud").Examples("http://localhost:8000", "http://127.0.0.1:3218")).
+		Field(service.NewStringField("workspace").Optional().Description("ID of the workspace. Required if target is `timeplus`.")).
+		Field(service.NewStringField("stream").Description("The name of the stream. Make sure the schema of the stream matches the input")).
+		Field(service.NewStringField("apikey").Secret().Optional().Description("The API key. Required if you are sending message to Timeplus Enterprise Cloud")).
+		Field(service.NewStringField("username").Optional().Description("The username. Required if you are sending message to Timeplus Enterprise (self-hosted) or timeplusd")).
+		Field(service.NewStringField("password").Secret().Optional().Description("The password. Required if you are sending message to Timeplus Enterprise (self-hosted) or timeplusd")).
+		Field(service.NewOutputMaxInFlightField()).
+		Field(service.NewBatchPolicyField("batching"))
+
+	if err := service.RegisterBatchOutput("timeplus", outputConfigSpec, newTimeplusOutput); err != nil {
+		panic(err)
+	}
+}
+
+type timeplus struct {
+	logger *service.Logger
+	client Writer
+}
+
+// Close implements service.Output
+func (t *timeplus) Close(context.Context) error {
+	// TODO: shall we wait for ongoing writes?
```

Comment:
I'd just get rid of `t.client = nil` here and, also the `t.client == nil` check in `Connect`. You'd get a panic if `service.RegisterBatchOutput` fails, so `t.client` can't be `nil`.

Reply:
updated.